data/tutorials/lg_08_error_handling.md

```diff
@@ -125,67 +191,549 @@ let () =
+(** @raise Not_found *)
+val find: ('a -> bool) -> 'a list -> 'a option
+```
+### Composing Functions Returning Options
+
+The function `div_opt` can't raise exceptions. However, since it doesn't return
+a result of type `int`, it can't be used in place of an `int`. The same way
+OCaml doesn't
+[promote](https://en.wikipedia.org/wiki/Type_conversion#Type_promotion) integers
+into floats, it doesn't automatically converts `int option` into `int` or _vice
+versa_.
+
+```ocaml
+# 21 + Some 21;;
+Error: This expression has type 'a option
+       but an expression was expected of type int
+```
+
+In order to combine option values with other values, conversion functions are
+needed. Here are the functions provided by the `Option` module to extract the
+data contained in an option:
+```ocaml
+val get : 'a t -> 'a
+val value : 'a t -> default:'a -> 'a
+val fold : none:'a -> some:('b -> 'a) -> 'b t -> 'a
+```
+`get` returns the content or raises `Invalid_argument` if applied to `None`.
+`value` essentially behaves as `get`, except it must be called with a default
+value which will be returned of if applied to `None`. `fold` also needs to be
+passed a default value that is returned when called on `None`, but it also
+expects a function that will be applied to the content of the option, when not
+empty.
+
+As a remark, observe that `value` can be implemented using `fold`:
+```ocaml
+# let value ~default = Option.fold ~none:default ~some:Fun.id;;
+val value : default:'a -> 'a option -> 'a = <fun>
+# Option.value ~default:() None = value ~default:() None;;
+- : bool = true
+# Option.value ~default:() (Some ()) = value ~default:() (Some ());;
+- : bool = true
+```
+
+It is also possible to perform pattern matching on option values:
+```ocaml
+match opt with
+| None -> ...    (* Something *)
+| Some x -> ...  (* Something else *)
+```
+However, sequencing such expressions leads to deep nesting which is often
+considered bad:
+
+> if you need more than 3 levels of indentation, you're screwed anyway, and
+> should fix your program.
+
+[Linux Kernel Style
+Guide](https://www.kernel.org/doc/Documentation/process/coding-style.rst)
+
+The recomended way to avoid that is to refrain from or delay attempting to
+access the content of an option value.
+
+### Using on `Option.map` and `Option.bind`
+
+Let's start with an example. Let's imagine one needs to write a function
+returning the [hostname](https://en.wikipedia.org/wiki/Hostname) part of an
+email address. For instance, given the email
+"gaston.lagaffe@courrier.dupuis.be," it would return "courrier."
+
+Here is a questionable but straightforward implementation using exceptions:
+```ocaml
+let host email =
+  let fqdn_pos = String.index email '@' + 1 in
+  let fqdn_len = String.length email - fqdn_pos in
+  let fqdn = String.sub email fqdn_pos fqdn_len in
+  try
+    let host_len = String.index fqdn '.' in
+    String.sub fqdn 0 host_len
+  with Not_found ->
+    if fqdn <> "" then fqdn else raise Not_found
```

Comment:
If I put the `;;` at the end, I get this output: `val host : string -> string = <fun>`